opflexagent/gbp_ovs_agent.py

```diff
@@ -277,15 +277,70 @@ def port_bound(self, port, net_uuid,
+                mapping_copy['floating_ip'] = []
+                fip_by_fixed = {}
+                original_mac = mapping.get('mac_address') or port.vif_mac
+                # Prepare FIPs by fixed_ip
+                for fip in mapping.get('floating_ip', []):
+                    fip_by_fixed.setdefault(
+                        fip['fixed_ip_address'], []).append(fip)
+                # For the main MAC, set floating IP collection to all those
+                # FIPs pointing to the active Port fixed ips.
+                for fixed in fixed_ips:
+                    mapping_copy['floating_ip'].extend(
+                        fip_by_fixed.get(fixed['ip_address'], []))
+
+                # For the main MAC EP, set al the AAP with no mac address or
+                # MAC address equal to the original MAC.
+                for aap in mapping.get('allowed_address_pairs', []):
+                    if aap.get('active'):
+                        if not aap.get('mac_address') or aap.get(
+                                'mac_address') == original_mac:
+                            # Should go with the MAIN mac address EP file
+                            mapping_copy['allowed_address_pairs'].append(aap)
+                            # Also set the right floating IPs
+                            mapping_copy['floating_ip'].extend(
+                                fip_by_fixed.get(aap['ip_address'], []))
+                        else:
+                            # Store for future processing
+                            mac_active_aap.setdefault(
+                                aap['mac_address'], []).append(aap)
+                # Create mapping file for base MAC address
+                self.mapping_to_file(port, net_uuid, mapping_copy, fixed_ips,
                                      device_owner)
+                # Reset for AAP EP files
+                mapping_copy['allowed_address_pairs'] = []
+                mapping_copy['fixed_ips'] = []
+                mapping_copy['extra_ips'] = []
+                mapping_copy['subnets'] = []
+                mapping_copy['enable_dhcp_optimization'] = False
+                mapping_copy['enable_metadata_optimization'] = False
+                # Map to file based on the active AAP with a MAC address
+                for mac, aaps in mac_active_aap.iteritems():
+                    # Replace the MAC address with the new one
+                    mapping_copy['mac_address'] = mac
+                    # Extend the FIP list based on the allowed IPs
+                    mapping_copy['floating_ip'] = []
+                    mapping_copy['ip-address-mapping'] = []
+                    for aap in aaps:
+                        mapping_copy['floating_ip'].extend(fip_by_fixed.get(
+                            aap['ip_address'], []))
+                    # For this mac, set all the allowed address pairs.
+                    mapping_copy['allowed_address_pairs'] = aaps
```

Comment:
Sorry I wasn't clear with my question:

Given that so far allowed_address_pairs only ended up filling the virtual-ip section... How did the port filtering work? How could a VM even send traffic from a AAP if the said address was never set in the "ip" section? Was it broken or is the opflex agent actually doing port security also considering the virtual-ip section?

As for the additional files, I believe that one or multiple files with their own virtual-ip sections should work seamlessly. The notion of "main" EP file should only exist in Openstack (artifact of a Neutron concept). Also consider that the virtual-ip sections of various EP files will never ovarlap, since they are always clustered by MAC address.